lib/CodeGen/CGDynamicCheck.cpp

```diff
@@ -111,6 +111,76 @@ void CodeGenFunction::EmitDynamicBoundsCheck(const Address PtrAddr, const Bounds
   EmitDynamicCheckBlocks(Builder.CreateAnd(LowerChk, UpperChk, "_Dynamic_check.range"));
 }
 
+void CodeGenFunction::EmitDynamicBoundsCheck(const Address BaseAddr,
+                                             const BoundsExpr *CastBounds,
+                                             const BoundsExpr *SubExprBounds) {
+  if (!getLangOpts().CheckedC)
+    return;
+
+  if (!SubExprBounds || !CastBounds)
+    return;
+
+  if (SubExprBounds->isAny() || SubExprBounds->isInvalid())
+    return;
+
+  // SubExprBounds can be Any by inference but CastBounds can't be Any
+  assert(!CastBounds->isAny());
+  if (CastBounds->isInvalid())
+    return;
+
+  // We can only generate the check if we have the bounds as a range.
+  if (!isa<RangeBoundsExpr>(SubExprBounds) ||
+      !isa<RangeBoundsExpr>(CastBounds)) {
+    llvm_unreachable(
+        "Can Only Emit Dynamic Bounds Check For RangeBounds Exprs");
+    return;
+  }
+
+  const RangeBoundsExpr *SubRange = dyn_cast<RangeBoundsExpr>(SubExprBounds);
+  const RangeBoundsExpr *CastRange = dyn_cast<RangeBoundsExpr>(CastBounds);
+
+  ++NumDynamicChecksRange;
+
+  // SubRange - bounds(lb, ub) vs CastRange - bounds(castlb, castub)
+  // Dynamic_check(lb <= castlb && castub <= ub)
+
+  // Emit the code to generate the pointer values
+  Address Lower = EmitPointerWithAlignment(SubRange->getLowerExpr());
+  Address Upper = EmitPointerWithAlignment(SubRange->getUpperExpr());
+
+  Value *LowerInt = Builder.CreatePtrToInt(Lower.getPointer(), IntPtrTy,
+                                           "_Dynamic_check.lower");
+  Value *UpperInt = Builder.CreatePtrToInt(Upper.getPointer(), IntPtrTy,
+                                           "_Dynamic_check.upper");
+
+  Address CastLower = EmitPointerWithAlignment(CastRange->getLowerExpr());
+  Address CastUpper = EmitPointerWithAlignment(CastRange->getUpperExpr());
+
+  Value *CastLowerInt = Builder.CreatePtrToInt(CastLower.getPointer(), IntPtrTy,
+                                               "_Dynamic_check.castlower");
+  Value *CastUpperInt = Builder.CreatePtrToInt(CastUpper.getPointer(), IntPtrTy,
+                                               "_Dynamic_check.castupper");
+
+  // Make the lower check (Lower <= CastLower)
+  Value *LowerChk =
+      Builder.CreateICmpULE(LowerInt, CastLowerInt, "_Dynamic_check.lower_cmp");
+
+  // Make the upper check (CastUpper <= Upper)
+  Value *UpperChk =
+      Builder.CreateICmpULE(CastUpperInt, UpperInt, "_Dynamic_check.upper_cmp");
+
+  // if expression E is NULL, skip dynamic_check
+  // Dynamic_check(E == NULL || (lb <= castlb && castub <= ub))
+  // if (cond1) {...}
+  // else {
+  // if (cond2) {...}
+  // else { trap(); llvm_unreachable(); }
+  // }
+  EmitDynamicCheckBlocks(
+      Builder.CreateIsNull(BaseAddr.getPointer(), "_Dynamic_check.null"),
```

Comment:
I think you should check for the non-null case before computing the upper and lower bound comparisons.   This doesn't quite match the semantics of the check.  In C, `(e1 || e2)` means that we first evaluate `e1`.    If that fails, then we evaluate `e2'.     In your implementation, we are always evaluating `e1` and `e2`.   Your implementation corresponds to `tmp1 = e1; tmp2 = e2`; if (tmp1 || tmp2)`.

The problem with always evaluating `e2' is that the upper/lower bound comparisons could end up doing pointer arithmetic involving null.   We are going to add runtime checks that disallow pointer arithmetic involving null.